src/control/cmd/dmg/auto.go

```diff
@@ -69,8 +69,12 @@ func (cmd *configGenCmd) Execute(_ []string) error {
 	}
 
 	resp, err := control.ConfigGenerate(ctx, req)
+	if err != nil {
+		// includes hardware validation errors e.g. hardware across hostset differs
+		return err
+	}
 
-	if resp != nil && resp.Errors() != nil {
```

Comment:
this change breaks behaviour, the reason for the weird process of deferring err check is that we want to print response errors even in the case that err is non-nil

Reply:
The way the code is currently written, Coverity complains because there is a check here to see if resp is not nil, but then down at line 87 there isn't a check and so it's possible that there could be a nil pointer dereference.

> we want to print response errors even in the case that err is non-nil

This is pretty confusing and is an anti-pattern. Functions should not return a valid response and an error. I will fix `ConfigGenerate()` to return one or the other, not both.

Reply:
Argh. I guess we do this elsewhere, too. Sigh. We really should fix that though. Instead of HostErrorsResp embedded in the response structs we should be returning a structured error that implements the `error` interface but can be cast into a richer data type for the pretty printers.

Reply:
I really should just read code first... I'm going to go back to my original position that Control API functions shouldn't be returning a non-nil resp _and_ a non-nil error. We don't actually do that elsewhere. The philosophy is that if anything goes wrong with invoking the RPCs, then we return an error. If something goes wrong on one of the target hosts, then we don't return an error but we do return a response with a populated HostErrorsMap.

Reply:
In the case of `ConfigGenerate()`, either the call succeeded and the response contains a generated configuration, or it failed, and the response is nil with a non-nil error. I've add a `config.ConfigGenerateError` type that implements `error` but can also be supplied to `pretty.PrintResponseErrors()` for more details. This keeps the API boring. Surprises are often nice, but not when they introduce inconsistent API semantics.

Reply:
I agree with the last comment and that's what we discussed in relation to the json handling in dmg and what I have implemented in https://github.com/daos-stack/daos/pull/5330. Re-factoring shouldn't be done in this PR though.

Reply:
OK, I've removed the fix for CID 329566 and put that issue in DAOS-7194 for you to fix, as you are more familiar with the code.